src/flick/show/views.py

```diff
@@ -27,54 +36,72 @@ class ShowViewSet(mixins.ListModelMixin, mixins.RetrieveModelMixin, viewsets.Gen
     # if api_settings.UNPROTECTED, then any user can see this
     permission_classes = api_settings.STANDARD_PERMISSIONS
 
-    # don't need this, generics has this code, but this overrides
-    # gives option to add additional checks / customize
-    # def list(self, request):
-    #     # can access logged in user via request.user
-    #     self.serializer_class = ShowSerializer
-    #     return super(ItemList, self).list(request)
 
-    # def retrieve(self, request, pk):
-    #     queryset = self.get_object()
-    #     serializer = ShowDetailSerializer(queryset, many=False)
-    #     return success_response(serializer.data)
+class SearchShow(APIView):
+    shows = []
+    known_shows = []
 
+    # show_type can be "movie", "tv", "anime"
+    def get_shows_by_type_and_query(self, query, show_type, source):
+        show_ids = local_cache.get((query, show_type))
+        if not show_ids:
+            show_ids = API.search_show_ids_by_name(show_type, query)
+            local_cache.set((query, show_type), show_ids)
+        for show_id in show_ids:
+            known_show = Show.objects.filter(ext_api_id=show_id, ext_api_source=source)
+            if known_show.exists():
+                known_show = Show.objects.get(ext_api_id=show_id, ext_api_source=source)
+                self.known_shows.append(ShowSerializer(known_show).data)
+            else:
+                show = API.get_show_info_from_id(show_type, show_id)
+                if show:
+                    self.shows.append(show)
 
-class SearchShow(APIView):
+    def get_shows_by_query(self, query, is_movie, is_tv, is_anime):
+        if is_movie:
+            self.get_shows_by_type_and_query(query, "movie", "tmdb")
+        if is_tv:
+            self.get_shows_by_type_and_query(query, "tv", "tmdb")
+        if is_anime:
+            self.get_shows_by_type_and_query(query, "anime", "animelist")
 
-    permission_classes = api_settings.UNPROTECTED
+    def get_top_shows_by_type(self, show_type):
+        top_shows = local_cache.get(("top", show_type))
```

Comment:
why are we searching in API only when there is nothing in cache? shows are quite time-sensitive, so top shows at the time of the first search and thus stored in cache might not always be the applicable "top show" at the time of the next search

Reply:
Also we should take into account new shows/updates on existing TV shows (new season/episode), but they may not be reflected if we just get from cache

Reply:
you're right 🥇 but shows don't change faster than items in our cache expires -- we can look into the optimal time expiration for when we should be fetching new things again